man/tcpclient.1
.TH tcpclient 1
.SH NAME
tcpclient \- create an outgoing TCP connection

.SH SYNOPSIS
.B tcpclient
[
.B \-46hHrRdDqQv
]
[
.B \-i\fIlocalip
]
[
.B \-p\fIlocalport
]
[
.B \-T\fItimeoutconn
]
[
.B \-l\fIlocalname
]
[
.B \-t\fItimeoutinfo
]
[
.B \-I\fIifname
]
.I host
.I port
.I program
[
.I arg ...
]

.SH DESCRIPTION
.B tcpclient
attempts to connect to a TCP server.
If it is successful, it runs
.I program
with the given arguments,
with descriptor 6 reading from the network
and descriptor 7 writing to the network.

The server's address is given by
.I host
and
.IR port .
.I host
may be 
.IR 0 ,
which is identical to
.IR 127.0.0.1 , 
or
.IR ::1
referring to the IPv4/IPv6 loopback address,
a compactified IPv6 address,
a dotted-decimal IPv4 address,
or a host name;
if a host has several IP addresses,
.B tcpclient
tries each in turn.
.I port
may be a numeric port number
or a port name.

.B tcpclient
sets up several environment variables,
as described in
.B tcp-environ(5).

.SH OPTIONS
.TP
.B \-4
Fall back to IPv4 sockets.  
.TP
.B \-6
Force IPv6 mode in UCSPI environment variables, even for
IPv4 connections.  This will set \fIPROTO\fR to \fBTCP6\fR and put
IPv4-mapped IPv6 addresses in \fITCPLOCALIP\fR and \fITCPREMOTEIP\fR.
.TP
.B \-i\fIlocalip
Use
.I localip
as the IP address for the local side of the connection;
quit if
.I localip
is not available.
.TP
.B \-p\fIlocalport
Use
.I localport
as the port number for the local side of the connection;
quit if
.I localport
is not available.
.TP
.B \-I\fIifname
Use
.I ifname 
as the local network interface.  This is only defined for IPv6 sockets
and needed if you use link-local IPv6 addresses.
.TP
.B \-T\fItimeoutconn
Give up on the
connection attempt
after
.I timeoutconn
seconds. Default: 60.
This timeout applies to each IP address tried.
.TP
.B \-d
(Default.)
Delay sending data for a fraction of a second whenever the
remote host is responding slowly,
to make better use of the network.
.TP
.B \-D
Never delay sending data;
enable TCP_NODELAY.
This is appropriate for interactive connections.
.TP
.B \-q
Quiet.
Do not print any messages.
.TP
.B \-Q
(Default.)
Print error messages.
.TP
.B \-v
Verbose.
Print all available messages.

.SH "DATA-GATHERING OPTIONS"
.TP
.B \-h
(Default.)
Look up the remote host name for
.IR TCPREMOTEHOST .
.TP
.B \-H
Do not look up the remote host name;
unset
.IR TCPREMOTEHOST .
.TP
.B \-l\fIlocalname
Do not look up the local host name;
use
.I localname
for
.IR TCPLOCALHOST .
.TP
.B \-r
(Default.)
Attempt to obtain
.I TCPREMOTEINFO
from the remote host.
.TP
.B \-R
Do not attempt to obtain
.I TCPREMOTEINFO
from the remote host.
.TP
.B \-t\fItimeoutinfo
Give up on the 
.I TCPREMOTEINFO
connection attempt
after
.I timeoutinfo
seconds. Default: 26.

.SH "ENVIRONMENT VARIABLES READ"
.B tcpclient
reads the environment variable
.I DNSCACHEIP
which can be used to specify the
recursive DNS Resolver dynamically.

.SH "SEE ALSO"
date@(1),
finger@(1),
http@(1),
mconnect(1),
tcpcat(1),
tcpserver(1),
who@(1),
tcp-environ(5)


doc/CHANGES
19960311
	version: ucspi-tcp 0.50, alpha.
19960803
	version: ucspi-tcp 0.60, alpha.
19960903
	version: ucspi-tcp 0.70, beta.
19960914
	version: ucspi-tcp 0.71, beta.
19960922
	version: ucspi-tcp 0.72, beta.
19970410
	version: ucspi-tcp 0.73, beta.
19980118
	version: ucspi-tcp 0.80, beta.
19981107
	code: switched to new install system.
	code: eliminated tcpclient -U; moved usage() to default.
	code: switched to strerr in tcpclient.
	code: switched to str_equal in tcpclient.
	portability problem: OpenBSD connect() prohibits 0.0.0.0.
		impact: tcpclient 0 fails. fix: interpret 0 as
		127.0.0.1. hopefully nobody needs 0 to mean primary
		interface.
	code: tcpclient now interprets empty host name as 0.
	doc: added addcr.0, delcr.0.
	doc: added who@.0, date@.0, finger@.0.
	doc: added tcpcat.0, mconnect.0.
	code: added printstatus() in tcpserver.
	code: eliminated numchildren in printpid() in tcpserver.
	code: unblocked SIGCHLD after setting SIGCHLD to SIG_DFL in
		tcpserver, instead of before, to avoid confusion if
		someone sends SIGCHLD manually.
	code: moved printstatus() before fork in tcpserver.
	code: added mconnect-io.
	code: added argv0.
	code: added fixcr.
	doc: revamped BLURB.
	code: added recordio.
	doc: cleaned up tcpclient.0.
	code: added http@.
	code: stopped after 100 bytes in safeappend() in tcpserver.
19981108
	code: added tcpserver -B.
	code: added a few regression tests.
	version: ucspi-tcp 0.83, beta.
19981109
	doc: added http@ to BLURB.
19981110
	code: added exec to mconnect.
	code: tcpclient now ignores port results from getpeername() in
		favor of the port it tried connecting to.
	code: added strport to tcpclient.
	doc: reorganized tcpclient.0.
	doc: reorganized tcpserver.0.
	code: revamped tcpserver messages.
	code: revamped tcpclient messages.
	code: added a few more regression tests.
	code: rewrote recordio for instant output and other features.
	doc: added references in tcpserver.0 and tcpclient.0.
	code: added tcpclient -i option.
	code: added tcprulescheck.
	code: added some tcprules regression tests.
19981111
	code: avoided ndelay in recordio.
	version: ucspi-tcp 0.84, beta.
19981212
	doc: improved description of host 0 in tcpserver.0. tnx to
		several people.
19981218
	doc: reordered sample rules in tcprules.0.
		tnx Harald Hanche-Olsen.
20000120
	internal: revamped packaging.
	internal: massive rewrite. tnx to everyone for bug reports and
		other comments on various versions of tcpclient,
		tcpserver, tcpcontrol, and rblsmtpd: Akihiro Iijima,
		Akihiro Terasaki, Albert J. deVera, Alex Vostrikov,
		Amos Shapira, Anand R. Buddhdev, Andrea Paolini,
		Andrew Pam, Araki Yasuhiro, Arne Wichmann,
		Ayamura Kikuchi, Bart B. B. Hanssens, Bert Gijsbers,
		Bradford M. Shelton, Brendan Kehoe, Brian J. Reichert,
		Brian M. Fisk, Brian O'Reilly, Bruno Wolff,
		Chris Garrigues, Chris Johnson, Christian Wettergren,
		Dale N. Woolridge, Dan M. Vogel, Daniel C. Mahoney,
		Darren Hall, Darren W. Rees, Dave Sill, David J. Walton,
		David P. Smith, David Pool, Dax Kelson, Dirk Jaeckel,
		Dirk Vluegels, Donald E. Blais, Ed Weinberg,
		Edward S. Marshall, Eric A. Perlman, Erik Wallin,
		Faried Nawaz, Fred B. Ringel, Frederik P. Lindberg,
		Gerry Boudreaux, Giles Lean, Grant Holliday,
		Greg D. Patterson, Harald Barth, Harald Fritzsche,
		Harald Hanche-Olsen, Hirofumi Ukawa, Hiroshi Yamashita,
		Hitesh Patel, Ingmar Hupp, J. B. Keith Humphreys,
		Jakub K. Boguslaw, Janos Farkas, Jason R. Mastaler,
		Jeff Hayward, Jeremy Wohl, Jim Littlefield,
		Johan Holmberg, John Bolhuis, John D. Mitchell,
		Jos Backus, Jose Monteiro, Joshua J. Ellis,
		Julie L. Baumler, Karl Lehenbauer, Karsten Thygesen,
		Keith Burdis, Kenny Elliott, Kikuchi Kousuke,
		Kris Kennaway, Lars Balker Rasmussen, Louis S. Theran,
		Lyndon F. Bartels, Mads E. Eilertsen, Mark Delany,
		Martin Mersberger, Matt P. Simerson, Matthew A. Zahorik,
		Michael B. Scher, Michael Handler, Michael Hirohama,
		Michael R. Gile, Michael Salmon, Mikael Suokas,
		Nobuhiro Murata, Patrick M. Kane, Paul R. Rotering,
		Peter Rye, Peter Wilkinson, Petr Novotny,
		Petri Kaukasoina, Raul D. Miller, Richard A. Soderberg,
		Robert W. Luce, Roberto A. Lumbreras, Russ Allbery,
		Russell Nelson, Sean Reifschneider, Shawn A. Clifford,
		Shin Ohira, Shinya O'Hira, Stan Norton,
		Stathy G. Touloumis, Stefan M. Linnemann, Sudish Joseph,
		Thomas E. Erskine, Thomas Kuerten, Tim Goodwin,
		Timothy L. Mayo, Todd Underwood, Tomoaki Terazawa,
		Tomoki Yoshioka, Toshinori Maeno, Uwe Ohse, Vern Hart,
		Vince Vielhaber, Waskita Adijarto, William E. Baxter,
		Wolfgang Rufeger, Wu Ching-hong, Yoshitatsu Takeshita.
20000307
	internal: switched to various new libraries.
	ui: tcpserver prints status on startup, before first connection.
20000309
	ui: incorporated rblsmtpd.
	ui: tcpserver -X allows nonexistent rules file.
	ui: http@ uses HTTP/1.0, sends Host, removes header.
	ui: added fixcrio.
	ui: delcr leaves CR alone at the end of a partial final line.
20000311
	ui: switched to prot; so setgid() is preceded by setgroups().
	ui: tcpserver supports -U.
	ui: tcpserver supports hostname rules.
	ui: tcprulescheck now uses environment variables.
20000312
	version: ucspi-tcp 0.86, beta.
20000314
	portability problem: the poll() emulation in RedHat 5.1 doesn't
		clear revents when select() returns 0. tnx Petr Novotny.
		impact: dns lookups busy-loop when they should time out.
		fix: new iopause from DNScache.
20000315
	version: ucspi-tcp 0.87, beta.
20000318
	internal: split rules() out of tcpserver and tcprulescheck.
	bug: didn't always clear rule length when $TCPREMOTEHOST was
		set. impact: empty rule could be ignored. fix: obvious.
		tnx Toshinori Maeno, Takashi Takizawa, Yuichi Katoh.
	ui: check for lone = rule if $TCPREMOTEHOST is set.
	ui: tcpclient supports fast+slow timeouts, 2+58 by default.
	ui: tcpclient randomizes order of addresses.
	version: ucspi-tcp 0.88, beta.
20080227 (feh)
	added Greetdelay option into rblsmtpd.

src/tcpclient.c
#include <unistd.h>
#include <sys/param.h>
#include <netdb.h>
#include <netinet/in.h>
#include "sig.h"
#include "exit.h"
#include "getoptb.h"
#include "uint_t.h"
#include "fmt.h"
#include "scan.h"
#include "str.h"
#include "ip.h"
#include "socket_if.h"
#include "fd.h"
#include "stralloc.h"
#include "buffer.h"
#include "error.h"
#include "logmsg.h"
#include "pathexec.h"
#include "timeout.h"
#include "timeoutconn.h"
#include "dnsresolv.h"
#include "byte.h"
#include "remoteinfo.h"

#define WHO "tcpclient"

void nomem(void)
{
  logmsg(WHO,111,FATAL,"out of memory");
}
void usage(void)
{
  logmsg(WHO,100,USAGE,"tcpclient \
[ -46hHrRdDqQv ] \
[ -i iplocal ] \
[ -p portlocal ] \
[ -T timeoutconn ] \
[ -l localname ] \
[ -t timeoutinfo ] \
[ -I interface ] \
host port program");
}

int verbosity = 1;
int flagdelay = 1;
int flagremoteinfo = 1;
int flagremotehost = 1;
unsigned long itimeout = 26;
unsigned long ctimeout[2] = { 2, 58 };
uint32 netif = 0;
int ipv4socket = 0;

char iplocal[16] = { 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0 };
uint16 portlocal = 0;
char *forcelocal = 0;

char ipremote[16];
uint16 portremote;

const char *hostname;
const char *loopback = "127.0.0.1";
static stralloc addresses;
static stralloc moreaddresses;

static stralloc tmp;
static stralloc fqdn;
char strnum[FMT_ULONG];
char ipstr[IP6_FMT];

char seed[128];

int main(int argc,char **argv)
{
  unsigned long u;
  int opt;
  char *x;
  int j;
  int s;
  int cloop;

  dns_random_init(seed);

  close(6);
  close(7);
  sig_ignore(sig_pipe);

  while ((opt = getopt(argc,argv,"46dDvqQhHrRi:p:t:T:l:I:")) != opteof)
    switch(opt) {
      case '4': ipv4socket = 1; break;
      case '6': ipv4socket = 0; break;
      case 'd': flagdelay = 1; break;
      case 'D': flagdelay = 0; break;
      case 'v': verbosity = 2; break;
      case 'q': verbosity = 0; break;
      case 'Q': verbosity = 1; break;
      case 'l': forcelocal = optarg; break;
      case 'H': flagremotehost = 0; break;
      case 'h': flagremotehost = 1; break;
      case 'R': flagremoteinfo = 0; break;
      case 'r': flagremoteinfo = 1; break;
      case 't': scan_ulong(optarg,&itimeout); break;
      case 'T': j = scan_ulong(optarg,&ctimeout[0]);
                if (optarg[j] == '+') ++j;
                scan_ulong(optarg + j,&ctimeout[1]);
                break;
      case 'i': if (!ip6_scan(optarg,iplocal)) usage(); break;
      case 'I': netif = socket_getifidx(optarg); break;
      case 'p': scan_ulong(optarg,&u); portlocal = u; break;
      default: usage();
    }
  argv += optind;

  if (!verbosity)
    buffer_2->fd = -1;

  hostname = *argv;
  if (!hostname || str_equal((char *)hostname,"")) usage();
  if (str_equal((char *)hostname,"0")) hostname = loopback;

  x = *++argv;
  if (!x) usage();
  if (!x[scan_ulong(x,&u)])
    portremote = u;
  else {
    struct servent *se;
    se = getservbyname(x,"tcp");
    if (!se)
      logmsg(WHO,111,FATAL,B("unable to figure out port number for: ",x));
    uint16_unpack_big((char *)&se->s_port,&portremote);
    /* i continue to be amazed at the stupidity of the s_port interface */
  }

  if (!*++argv) usage();

  if (ip4_scan(hostname,ipremote + 12)) {
    if (!stralloc_copys(&addresses,"")) nomem();
    byte_copy(addresses.s,12,V4mappedprefix);
    byte_copy(addresses.s + 12,4,ipremote + 12);
    ipv4socket = 1;
  } else if (ip6_scan(hostname,ipremote))
      if (!stralloc_copyb(&addresses,ipremote,16)) nomem();

  if (!addresses.len) {
    if (!stralloc_copys(&tmp,hostname)) nomem();
    if (dns_ip6_qualify(&addresses,&fqdn,&tmp) < 0)
      logmsg(WHO,111,FATAL,B("unable to figure out IP address for: ",(char *)hostname));
  }
  if (addresses.len < 16) 
    logmsg(WHO,111,FATAL,B("no IP address for: ",(char *)hostname));

  if (addresses.len == 16) {
     ctimeout[0] += ctimeout[1];
     ctimeout[1] = 0;
  }

  for (cloop = 0; cloop < 2; ++cloop) {
    if (!stralloc_copys(&moreaddresses,"")) nomem();
    for (j = 0; j + 16 <= addresses.len; j += 16) {
      ipv4socket =  ip6_isv4mapped(addresses.s + j);
      s = socket_tcp();
      if (s == -1)
        logmsg(WHO,111,FATAL,"unable to create socket");
      if (socket_bind(s,iplocal,portlocal,netif) == -1)
        logmsg(WHO,111,FATAL,"unable to bind socket");
      if (timeoutconn(s,addresses.s + j,portremote,ctimeout[cloop],netif) == 0)
        goto CONNECTED;
      close(s);
      if (!cloop && ctimeout[1] && (errno == ETIMEDOUT)) {
        if (!stralloc_catb(&moreaddresses,addresses.s + j,16)) nomem();
      }
      else {
        strnum[fmt_ulong(strnum,portremote)] = 0;
        if (ip6_isv4mapped(addresses.s + j))
          ipstr[ip4_fmt(ipstr,addresses.s + j + 12)] = 0;
        else
          ipstr[ip6_fmt(ipstr,addresses.s + j)] = 0;
      }
    }
    if (!stralloc_copy(&addresses,&moreaddresses)) nomem();
  }
  logmsg(WHO,-99,DROP,B("unable to connected to: ",ipstr," port: ",strnum));

  _exit(111);


  CONNECTED:

  if (!flagdelay)
    socket_tcpnodelay(s); /* if it fails, bummer */

  if (socket_local(s,iplocal,&portlocal,&netif) == -1)
    logmsg(WHO,111,FATAL,"unable to get local address");

  if (ip6_isv4mapped(iplocal) || byte_equal(iplocal,16,V6any))
    ipv4socket = 2;

  if (!pathexec_env("PROTO",ipv4socket?"TCP":"TCP6")) nomem();

  strnum[fmt_ulong(strnum,portlocal)] = 0;
  if (!pathexec_env("TCPLOCALPORT",strnum)) nomem();

  if (!ipv4socket) {
    ipstr[ip6_fmt(ipstr,iplocal)] = 0;
    if (!pathexec_env("TCP6LOCALIP",ipstr)) nomem();
    if (!pathexec_env("TCP6LOCALPORT",strnum)) nomem();
  } else
    ipstr[ip4_fmt(ipstr,iplocal + 12)] = 0;
  if (!pathexec_env("TCPLOCALIP",ipstr)) nomem();

  x = forcelocal;
  if (!x)
    if (dns_name6(&tmp,iplocal) < 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPLOCALHOST",x)) nomem();

  if (socket_remote(s,ipremote,&portremote,&netif) == -1)
    logmsg(WHO,111,FATAL,"unable to get remote address");

  strnum[fmt_ulong(strnum,portremote)] = 0;
  if (!pathexec_env("TCPREMOTEPORT",strnum)) nomem();

  if (!ipv4socket) {
    ipstr[ip6_fmt(ipstr,ipremote)] = 0;
    if (!pathexec_env("TCP6REMOTEIP",ipstr)) nomem();
    if (!pathexec_env("TCP6REMOTEPORT",strnum)) nomem();
  } else
    ipstr[ip4_fmt(ipstr,ipremote + 12)] = 0;
  if (!pathexec_env("TCPREMOTEIP",ipstr)) nomem();

  if (verbosity >= 2)
    log_who(WHO,B("connected to ",ipstr," port ",strnum));

  x = 0;
  if (flagremotehost)
    if (dns_name6(&tmp,ipremote) < 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPREMOTEHOST",x)) nomem();

  x = 0;
  if (flagremoteinfo)
    if (remoteinfo(&tmp,ipremote,portremote,iplocal,portlocal,itimeout,netif) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPREMOTEINFO",x)) nomem();

  if (fd_move(6,s) == -1)
    logmsg(WHO,111,FATAL,"unable to set up descriptor 6");
  if (fd_copy(7,6) == -1)
    logmsg(WHO,111,FATAL,"unable to set up descriptor 7");
  sig_uncatch(sig_pipe);

  pathexec(argv);
  logmsg(WHO,111,FATAL,B("unable to run: ",*argv));

  return 111;
}
